// @xliic/common/audit.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.groupNames = exports.CriticalityLevelName = exports.CriticalityLevelId = exports.Criticality = exports.SeverityLevels = void 0;
exports.SeverityLevels = ["info", "low", "medium", "high", "critical"];
exports.Criticality = {
    Info: 1,
    Low: 2,
    Medium: 3,
    High: 4,
    Critical: 5,
};
exports.CriticalityLevelId = {
    [exports.Criticality.Info]: "info",
    [exports.Criticality.Low]: "low",
    [exports.Criticality.Medium]: "medium",
    [exports.Criticality.High]: "high",
    [exports.Criticality.Critical]: "critical",
};
exports.CriticalityLevelName = {
    [exports.Criticality.Info]: "Info",
    [exports.Criticality.Low]: "Low",
    [exports.Criticality.Medium]: "Medium",
    [exports.Criticality.High]: "High",
    [exports.Criticality.Critical]: "Critical",
};
exports.groupNames = {
    oasconformance: {
        validation: "Structure",
        semantics: "Semantics",
        bestpractices: "Best practices",
    },
    datavalidation: {
        parameters: "Parameters",
        paths: "Paths",
        schema: "Schema",
        responseheader: "Response headers",
        responsedefinition: "Response definition",
    },
    security: {
        authentication: "Authentication",
        authorization: "Authorization",
        transport: "Transport",
    },
};
