// @xliic/common/oas30.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getServerUrls = exports.OasPrimitiveTypes = exports.getOperations = exports.getParametersMap = exports.getOperationParameters = exports.getPathItemParameters = exports.getOperation = void 0;
const http_1 = require("./http");
const ref_1 = require("./ref");
function getOperation(oas, path, method) {
    return (0, ref_1.deref)(oas, oas.paths[path])?.[method];
}
exports.getOperation = getOperation;
function getPathItemParameters(oas, pathItem) {
    const params = pathItem.parameters ?? [];
    return params.map((param) => (0, ref_1.deref)(oas, param));
}
exports.getPathItemParameters = getPathItemParameters;
function getOperationParameters(oas, operation) {
    const params = operation?.parameters ?? [];
    return params.map((param) => (0, ref_1.deref)(oas, param));
}
exports.getOperationParameters = getOperationParameters;
function getParametersMap(oas, pathParameters, operationParameters) {
    const result = { query: {}, header: {}, path: {}, cookie: {} };
    // path parameters first, to allow them to be overriden
    for (const parameter of pathParameters) {
        const schema = (0, ref_1.deref)(oas, parameter.schema);
        result[parameter.in][parameter.name] = { ...parameter, schema };
    }
    // potentially override path parameters using ones defined in the operation itself
    for (const parameter of operationParameters) {
        const schema = (0, ref_1.deref)(oas, parameter.schema);
        result[parameter.in][parameter.name] = { ...parameter, schema };
    }
    return result;
}
exports.getParametersMap = getParametersMap;
function getOperations(oas) {
    const operations = [];
    for (const path of Object.keys(oas.paths)) {
        for (const method of Object.keys(oas.paths[path])) {
            if (http_1.HttpMethods.includes(method)) {
                const operation = getOperation(oas, path, method);
                operations.push([path, method, operation]);
            }
        }
    }
    return operations;
}
exports.getOperations = getOperations;
exports.OasPrimitiveTypes = ["string", "number", "integer", "boolean"];
function getServerUrls(oas) {
    const servers = (oas.servers ?? [])
        .filter((server) => server.url !== undefined && server.url !== "")
        .map((server) => server.url);
    if (servers.length > 0) {
        return servers;
    }
    return ["http://localhost/"];
}
exports.getServerUrls = getServerUrls;
